src/generators/gradient_based/loss.jl

```diff
@@ -39,9 +37,15 @@ function ∇(
     M::Models.AbstractDifferentiableModel,
     ce::AbstractCounterfactualExplanation,
 )
-    ℓ = 0
-    if (ce.convergence[:converge_when] == :invalidation_rate)
-        ℓ = hinge_loss(ce)
-    end
-    return ∂ℓ(generator, M, ce) + ∂h(generator, ce) .+ ℓ
+    return ∂ℓ(generator, M, ce) + ∂h(generator, ce) .+ hinge_loss(ce.convergence, ce)
+end
+
+"""
+    hinge_loss(convergence::InvalidationRateConvergenc, ce::AbstractCounterfactualExplanation)
+
+The default hinge loss for any convergence criterion.
+Can be overwritten inside the `convergence` folder as part of the definition of specific convergence criteria.
+"""
+function hinge_loss(convergence::AbstractConvergence, ce::AbstractCounterfactualExplanation)
+    return 0
 end
```

Comment:
I created this method because defining a default `hinge_loss()` method inside the `Generators` module and overriding it inside the `Convergence` module for the `InvalidationRateConvergence` criterion seemed to be the only way of avoiding cyclical dependencies between the `Convergence` and `Generators` modules. If I had defined `hinge_loss` only inside the `Convergence` module, I would have had to call `Convergence.hinge_loss()` inside the `∇` method, while if I had defined `hinge_loss` only inside the `Generators` module, I would have had to call `Convergence.InvalidationRateConvergence` inside the `hinge_loss()` method. The `Convergence` module has to make calls to the `Generators` module anyways, so there's no way to allow `Convergence` to be a dependency of `Generators`. I'm aware that this solution is not ideal, though, so if anyone has better ideas, let me know.